### DISCUSSION TITLE
Processor Specific Decompiler Rules

### DISCUSSION THREAD

**Discussion by astrelsky** [NationalSecurityAgency/ghidra#2991] [Ideas]
This could be done two ways.

1. Utilizing the  [rule compiler](https://www.github.com/NationalSecurityAgency/ghidra/tree/master/Ghidra%2FFeatures%2FDecompiler%2Fsrc%2Fdecompile%2Fcpp%2Frulecompile.cc) and having the rule file specified in in the `ldefs` file. The compiler does appear to be functional. However extreme care is required as it will happily dereference an uninitialized pointer. This does need a bit of work. Dealing with the uninitialized pointer dereference can be dealt with a custom "smart" pointer template class. This class would default it's value to `nullptr` and it would check prior to dereferencing to provide a proper explanation of what went wrong in the low level error. 

2. Via an exported vector of rules from a user created dynamic library.

This would help ease the burden of developing new rules and would also ease the pain of dealing with rules that only apply for specific processors.

**Comment by hedgeberg** [comment]
I'd really like to see something like this, and also something to allow extensions or even end users to add rules that could be enabled/disabled at runtime. One of ghidra's biggest weaknesses, currently, is the limited ability to do any kind of transformation on the output of the decompiler from the java UI. User-controlled rules that could be invoked at the end of the decompilation pipeline to do things like apply transformations to the resolved syntax tree would open up a whole new realm of possibilities for manipulating the output of the decompiler. I could even see it being useful for decompiling to things like non-C/C++ languages.

**Reply by fmagin** [reply]
I actually have a prototype plugin for this. The main limitation is that a lot of information is kept in the decompiler, and you basically either need a parser for the Clang Tokens in Java, or Java code that translates the PCode AST into Clang Tokens again (which usually happens in the decompiler). There isn't really a "syntax tree" that is available in the Java process.

**Reply by astrelsky** [reply]
> I actually have a prototype plugin for this. The main limitation is that a lot of information is kept in the decompiler, and you basically either need a parser for the Clang Tokens in Java, or Java code that translates the PCode AST into Clang Tokens again (which usually happens in the decompiler). There isn't really a "syntax tree" that is available in the Java process.

There is a custom grammar language for decompiler rules. I had built and enabled it after a few fixes but found that it was so easy to segfault from a nullptr dereference and because it is not documented at all I decided it wasn't worth the effort.

**Reply by fmagin** [reply]
Do you have a writeup of those fixes somewhere? I tried building the decompiler with `ADDITIONAL_FLAGS=-DCPUI_RULECOMPILE` yesterday, but didn't manage. Issues like ` undefined reference to 'vtable for RuleGeneric'`.

I'll try to find some time to clean up my plugin prototype and publish it. It requires a Ghidra fork with some minor changes, (one new service that is used in the DecompilerPlugin, and adding a bunch of constructors for the Clang Tokens, that part is already public https://github.com/fmagin/ghidra/tree/Ghidra_10.1.1_build-transformer ). But then any kind of code can just register a function that transforms a `DecompileResults` into a new `DecompileResults`, those are then each run on any `DecompileResults` returned by the decompiler binary. It is even possible to write a function in the Jython shell and register it.

**Reply by astrelsky** [reply]
I'll see if I still have it laying around. However that undefined reference is probably due to a class with declared virtual functions but none of them are defined. 

I just checked and I don't have it anymore. However looking at the class everything seems to be defined. It `rulecompile.cc` actually being compiled and included along with `ruleparse.cc` and `ruleparse.hh`?

**Reply by msm-code** [reply]
I just got around to blogging about this. If you're still interested in how this DSL works, check out https://msm.lt/re/ghidra/rulecompile/

**Reply by astrelsky** [reply]
> I just got around to blogging about this. If you're still interested in how this DSL works, check out https://msm.lt/re/ghidra/rulecompile/

I did not know it was possible to make the decompiler use them automatically when running in the UI. From what I can remember from a few years ago, the language seemed incomplete and it was very easy to cause the decompiler to segfault. There was no verification of the rules.

**Reply by msm-code** [reply]
It's still not possible, I solved this with a dirty hack - basically I hardcoded a path, and forced the decompiler to always load rules from there. Obviously not a proper solution, but I wanted to check if that really works.

I'm not sure about the completeness of the language (it's powerful enough to write many useful rules), but what you say about segfaults is **absolutely true**. When testing this, I witnessed dozens of segfaults. Doing anything wrong, for example a typo in a variable name, causes a segfault. Clearly it's not production ready - probably why it was disabled in the first place.

**Reply by astrelsky** [reply]
> It's still not possible, I solved this with a dirty hack - basically I hardcoded a path, and forced the decompiler to always load rules from there. Obviously not a proper solution, but I wanted to check if that really works.
> 
> I'm not sure about the completeness of the language (it's powerful enough to write many useful rules), but what you say about segfaults is **absolutely true**. When testing this, I witnessed dozens of segfaults. Doing anything wrong, for example a typo in a variable name, causes a segfault. Clearly it's not production ready - probably why it was disabled in the first place.

At this point it's practically abandonware. However, I could see it being used in it's current state for something like fuzzing to find flaws or bad assumptions about the state of varnodes in other decompiler rules.

It's just unfortunately because it is clear that someone put some thought and work into this at some point.

**Reply by msm-code** [reply]
Yeah. I would really, really, like to have a way to influence decompiler output (more than currently possible). I work with obfuscated code a lot, and that would be great capability to have. This discussion didn't much attention 3 years ago, but who knows, maybe 2025 will be the year of ghidra decompiler.

**Reply by osogi** [reply]
Hi, 2025 here :) 
I’m trying to do something similar to what’s discussed here. If you’ve had any new thoughts since this thread, or have any advice/pointers, I’d really appreciate it.

Link: https://github.com/NationalSecurityAgency/ghidra/discussions/8758
